webdriver/tests/interaction/element_click.py

```diff
@@ -0,0 +1,142 @@
+from tests.support.inline import inline
+
+def click(session, element):
+    return session.transport.send("POST", "session/{session_id}/element/{element_id}/click"
+                                  .format(session_id=session.session_id,
+                                          element_id=element.id))
+
+
+# 14.1 Element Click
+
+def test_no_browsing_context(session, create_window):
+    # 14.1 step 1
+    session.url = inline("<p>This is not an editable paragraph.")
+    element = session.find.css("p", all=False)
+
+    session.window_handle = create_window()
+    session.close()
+
+    response = click(session, element)
+    assert_error(response, "no such window")
+
+
+def test_element_file_upload_state(session):
+    # 14.1 Step 3
+    session.url = inline("<input type=file value=Federer>")
+
+    element = session.find.css("input", all=False)
+    response = click(session, element)
+    assert_error(response, "invalid argument")
+
+
+def test_scroll_into_element_view(session):
+    # 14.1 Step 4
+    session.url = inline("<input type=text value=Federer><div style= \"height: 200vh; width: 5000vh\">")
+
+    # Scroll to the bottom right of the page
+    session.execute_script("window.scrollTo(document.body.scrollWidth, document.body.scrollHeight);")
+    element = session.find.css("input", all=False)
+    # Clear and scroll back to the top of the page
+    response = click(session, element)
+    assert_success(response)
+
+    # Check if element cleared is scrolled into view
+    rect = session.execute_script("return document.getElementsByTagName(\"input\")[0].getBoundingClientRect()")
+
+    pageDict = {}
+
+    pageDict["innerHeight"] = session.execute_script("return window.innerHeight")
+    pageDict["innerWidth"] = session.execute_script("return window.innerWidth")
+    pageDict["pageXOffset"] = session.execute_script("return window.pageXOffset")
+    pageDict["pageYOffset"] = session.execute_script("return window.pageYOffset")
+
+    assert rect["top"] < (pageDict["innerHeight"] + pageDict["pageYOffset"]) and \
+           rect["left"] < (pageDict["innerWidth"] + pageDict["pageXOffset"]) and \
+           (rect["top"] + element.rect["height"]) > pageDict["pageYOffset"] and \
+           (rect["left"] + element.rect["width"]) > pageDict["pageXOffset"]
+
+# TODO
+# Any suggestions on implementation?
+# def test_element_not_interactable(session):
+#     # 14.1 Step 5
+#     assert_error(response, "element not interactable")
+
+def test_element_click_intercepted(session):
+    # 14.1 Step 6
+    session.url = inline("""<input type=button value=Roger style=\"position: absolute; left: 10px; top: 10px\">
+                            <div style=\"position: absolute; height: 100px; width: 100px; background: rgba(255,0,0,.5); left: 10px; top: 5px\"></div>""")
+
+    element = session.find.css("input", all=False)
+    response = click(session, element)
+    assert_error(response, "element not interactable")
+
+
```

Comment:
Please can you add a test where we click on a link that loads a new page with the following different types:

- click a link with just numbers
- click a link where its a multiline string
- unload event is fired
- link is just a hash (so doesn't change the page)
- closes the window/tab

Reply:
Please can you add a test where there is are CSS transforms:

- transformed off the page
- rotated

Reply:
Hey @AutomatedTester,

Could you explain the click a link with just numbers and multiline string test a little more? Is it a link with just numbers or a link that opens a page with just numbers? Could you provide an example? 

Thanks

Reply:
test with number https://dxr.mozilla.org/mozilla-central/source/testing/marionette/harness/marionette_harness/tests/unit/test_click.py#99

As for a multiline link click we are checking that we only work with the first element rect of the element not just the bounding client rect

Reply:
Hey @AutomatedTester, I used some of html elements in clicks.html from your link. Is that ok or should I make my elements different?

Thanks

Reply:
Also, could you give me some insight on clicking a link to close a window/tab? Looks like I can't do window.close() unless the window was opened by it. Were you thinking something like click a link to a webpage with a javascript to open and close that window and then check if that window still exist?

Reply:
You will have to open a window through `window.open(…)`, then have an event on a button in that document that closes itself, correct.

Reply:
fixed